Retrieve list in chunks in cronjob controller

jingyih opened the pull request:
**What this PR does / why we need it**:
Retrieve jobs and cronjobs in chunks.  This reduces server resource consumption, improves server responsiveness.

For example, we have seen the cronjob controller listing 320k jobs from the etcd server, which takes over 30 seconds to finish.
```
I1016 00:39:03.152718       1 trace.go:76] Trace[2046021670]: "List /apis/batch/v1/jobs" (started: 2018-10-16 00:38:29.832824845 +0000 UTC m=+380.058697937) (total time: 33.319846272s):
Trace[2046021670]: [21.225676154s] [21.225669191s] Listing from storage done
Trace[2046021670]: [21.57869171s] [353.015556ms] Self-linking done
Trace[2046021670]: [33.319842654s] [11.741150944s] Writing http response done (320186 items)
```

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig apps

yue9944882 commented:
/ok-to-test

justinsb commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -104,19 +106,35 @@ func (jm *CronJobController) syncAll() {
+	cronJobListFunc := func(opts metav1.ListOptions) (runtime.Object, error) {
+		return jm.kubeClient.BatchV1beta1().CronJobs(metav1.NamespaceAll).List(opts)
+	}
+	sjlTmp, err := pager.New(pager.SimplePageFunc(cronJobListFunc)).List(context.TODO(), metav1.ListOptions{})
```
Rename from `sjl`/`sjlTmp` -> `cjl`/`cjlTmp` (the `s` refers to the former name for cronjobs, I think)?  It's actually probably more readable just to spell it out though - e.g. `jobs`, `cronjobs`, `cronjobList` etc

jingyih replied:
Thanks! I was trying to minimize the change, but spelling out sounds good.

justinsb commented:
Have we benchmarked this to be a win?  I think pagination hits etcd directly, so it might be more resource intensive on both apiserver and etcd.

I do see that the memory hit of a list of 320k Jobs is going to be painful though... perhaps it is time for a GRPC interface for listing objects so we can easily stream them!

An alternative approach might be to list all cronjobs, then loop through each and fetch just the jobs with that owner.  (Can we filter by owner?)

Dare I ask how we got up to 320k jobs?  That seems like an awful lot :-)

soltysh commented:
/hold
I don't think this was discussed in the sig-apps anywhere or I missed it. During this Monday's meeting all of us agreed that instead of improving the current implementation we need to work to get #17130 solved. For that we should not complicate the current flow of the controller, which will only make the re-write more complicated. 

@kow3ns @janetkuo  agreed?

janetkuo commented:
> An alternative approach might be to list all cronjobs, then loop through each and fetch just the jobs with that owner. (Can we filter by owner?)

Unfortunately we cannot do that because Jobs are listed first intentionally to avoid race with GC controller. 
https://github.com/kubernetes/kubernetes/blob/117c6177c4aa34df04a8a84b5789d764a5b27525/pkg/controller/cronjob/cronjob_controller.go#L103-L107

cc @enisoc

janetkuo commented:
> During this Monday's meeting all of us agreed that instead of improving the current implementation we need to work to get #17130 solved. For that we should not complicate the current flow of the controller, which will only make the re-write more complicated.

Once we have #17130 we can list from cache and then we probably don't need "list in chunks" at all.

jingyih commented:
@justinsb The current code is listing from etcd directly.  The change I am proposing is merely adding pagination to listing.  The intention of the PR is to help resource consumption in apiserver and etcd, and to improve responsiveness.

The 320k jobs were created by customer.  We tried to reach out to them but so far got no response.  So the jobs are just sitting there, being listed by cronjob controller every 10 seconds...

jingyih commented:
@soltysh I agree with [#17130](https://github.com/kubernetes/kubernetes/issues/17130), listing from etcd directly could be expensive.  Before the re-write happens, this PR that I am proposing is merely try to make the situation a little better, by adding pagination to listing.  I do not think this change adds any complexity to the future re-write (with informer).

jpbetz commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -104,19 +106,35 @@ func (jm *CronJobController) syncAll() {
+	jobListFunc := func(opts metav1.ListOptions) (runtime.Object, error) {
+		return jm.kubeClient.BatchV1().Jobs(metav1.NamespaceAll).List(opts)
+	}
+	jlTmp, err := pager.New(pager.SimplePageFunc(jobListFunc)).List(context.TODO(), metav1.ListOptions{})
```
I believe `context.Background()` hints at our intent a bit better than `context.TODO()` here, although they are otherwise equivalent (I believe).

jingyih replied:
Done.

jingyih committed:
controller: use pagination in cronjob controller

Use pagination (chunking) when listing Jobs and CronJobs in cronjob
controller.

jpbetz commented:
/lgtm

Not sure why `pull-kubernetes-verify` is failing with the `import k8s.io/client-go/tools/pager did not match any allowed prefix` error. Rebase and see if the linter is misconfigured and has since been fixed?

jpbetz commented:
/lgtm

jpbetz commented:
@soltysh, @derekwaynecarr,, I see that a hold was put on this issue due to https://github.com/kubernetes/kubernetes/issues/17130.  I propose we go ahead with this fix since (1) it's fairly straightforward and doesn't preclude switching to informers in the future, (2) it makes things better in the near term.  Any concerns?

jingyih commented:
@soltysh With pagination, a single large list operation is divided into multiple lists of 500 items. 500 is the default page size, used by listwatcher, which is used by shared informer.  Trace is only enabled when a request is taking a long time.  In this case, a listing of 500 items will not trigger trace.  So I do not have a direct trace comparison.

Again, this change is just following the K8s API conventions: https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks

janetkuo reviewed (approved):
This PR allows CronJob controller to retrieve LIST results in pages to reduce memory impact, with this beta feature: https://github.com/kubernetes/enhancements/issues/365

This also doesn't complicate existing code/design. Given that the PR for switching to informer isn't ready yet, I'd +1 this.

soltysh reviewed (approved):
Ok, let's land this. We won't make things worse, for sure. Thanks!
/hold cancel
/approve

soltysh commented:
@janetkuo mind giving final approve?

janetkuo commented:
/approve

jpbetz commented:
/ok-to-test

tedyu commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -102,19 +104,35 @@ func (jm *CronJobController) syncAll() {
 	}
+	jl, ok := jlTmp.(*batchv1.JobList)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expected type *batchv1.JobList, got type %T", jlTmp))
```
Don't we need to use reflect to get the type of jlTmp ?

jpbetz replied:
Shouldn't need to, `%T` is documented as "a Go-syntax representation of the type of the value".
